### ISSUE TITLE
Regular expression virtual host

### ISSUE THREAD

**Issue by abdollahpour** [nginx-proxy/nginx-proxy#117]
I'm try to use regex for domain matching (*-staging.my-domain.com)

```
-e VIRTUAL_HOST="~^.*\-staging\.my\-domain\.com$"
```

Any idea why it does not work?

**Comment by md5** [comment]
I believe the issue is that the `upstream` is being named "`~^.*\-staging\.my\-domain\.com$`", so `nginx` is choking on it.

I could see a few different fixes:

1. Use the `sha` function in `docker-gen` to hash all `VIRTUAL_HOST` values before using them as an `upstream` name.
    * **Pros**: Trivially avoids all issues with special characters by using only hex strings.
    * **Cons**: Debugging will be nasty since it will be full of SHA hashes. Unclear how to handle naming for SSL certificates, htpasswd files, and custom configs.
2. Normalize `VIRTUAL_HOST` values that contain certain special characters for use as the `upstream` name, leaving normal characters intact.
    * **Pros**: May be difficult to get normalization correct.
    * **Cons**: Will have to be a new function in `docker-gen`. Normalization scheme will have to be documented and understandable.
3. Support directly naming the virtual host via a `VIRTUAL_NAME` or possibly by introducing a `VIRTUAL_HOST_CONFIG` JSON blob using the new `parseJson` functionality I added to `docker-gen`.
    * **Pros**: Allows direct user control of naming of `upstream`.
    * **Cons**: May need to be a prefix instead of direct naming since multi-host `VIRTUAL_HOST` setups are supported (e.g. `{{Env.VIRTUAL_NAME}}-{{sha $host}}`).

**Comment by danhealy** [comment]
What specifically causes it to fail?

I was trying to get this to work:

    ~^example\.com(\..*\.xip\.io)?$

and I had to settle for 

    example.*

**Comment by md5** [comment]
This is what I see:

```
$ docker exec nginx-proxy nginx -t
2015/03/07 18:52:00 [emerg] 95#0: invalid variable name in /etc/nginx/conf.d/default.conf:128
nginx: [emerg] invalid variable name in /etc/nginx/conf.d/default.conf:128
nginx: configuration file /etc/nginx/nginx.conf test failed
$ docker exec nginx-proxy sed -n '128p' /etc/nginx/conf.d/default.conf
		proxy_pass http://~^.*-staging\.my-domain\.com$;
```

@jwilder added #115 to track adding a call to `nginx -t` to check the config before reloading, but the feature to support that still needs to be implemented in [`jwilder/docker-gen`](https://github.com/jwilder/docker-gen).

As as I said in the other comment, the issue is that "~^.*-staging\.my-domain\.com$" is being used to name the `upstream` and that is not a valid Nginx variable name.

**Comment by md5** [comment]
I just realized that URL encoding is a good way to implement my suggestion 2, so I've opened jwilder/docker-gen#72 to expose the `QueryEscape` function from `net/url` in `docker-gen`.

**Comment by md5** [comment]
Looks like the merge in jwilder/docker-gen#72 closed this without it actually getting fixed. A call or two to `queryEscape` will need to be added to `nginx.tmpl`.

**Comment by buchdag** [comment]
fixed by #671
